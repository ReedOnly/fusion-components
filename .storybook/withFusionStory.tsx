import React from 'react';
import Header from '../src/components/core/Header';
import Content from '../src/components/core/Content';
import Overlay from '../src/components/general/ApplicationGuidance/components/Overlay';

export default (title: string, padding: number = 16) => (stories: () => React.ReactNode) => {
    return (
        <>
            <Overlay fixed>
                <Header aside={null} start={null} content={null} />
            </Overlay>
            <Content>
                <Overlay>
                    <div
                        style={{
                            padding,
                            minHeight: '100%',
                            width: '100%',
                            maxWidth: '100%',
                            boxSizing: 'border-box',
                        }}
                    >
                        {stories()}
                    </div>
                </Overlay>
            </Content>
        </>
    );
};
