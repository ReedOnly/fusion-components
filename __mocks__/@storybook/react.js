import renderer from 'react-test-renderer';
import addons, { mockChannel } from '@storybook/addons';
import { withFusionContext } from '../../.storybook/withFusionContext';

addons.setChannel(mockChannel());

export const action = (actionName) => jest.fn();
const fusionContext = withFusionContext();
export const storiesOf = (groupName) => {
    const api = {
        decorators: [],
        parameters: {},

        add(storyName, story) {
            if (api.parameters.skipTest) {
                return;
            }

            describe(groupName, () => {
                it(storyName, () => {
                    // const decoratedStory = api.decorators.reduce((s, decorator) => {
                    //     try {
                    //         return decorator(s, { parameters: api.parameters });
                    //     } catch (e) {
                    //         return s;
                    //     }
                    // }, story);

                    const component = renderer.create(fusionContext(story));

                    // @TODO Enable custom names once released:
                    // > https://github.com/facebook/jest/pull/2094
                    expect(component.toJSON())
                        .toMatchSnapshot
                        // `${groupName}.${storyName}`
                        ();
                });
            });

            return api;
        },

        addDecorator() {
            return api;
        },

        addParameters(parameters) {
            api.parameters = { ...api.parameters, ...parameters };
            return api;
        },
    };

    return api;
};

export const addDecorator = () => {};
