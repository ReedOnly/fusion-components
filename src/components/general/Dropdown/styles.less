.dropdownContainer {
    position: absolute;
    top: 100%;
    min-width: 100%;
    margin-top: calc(var(--grid-unit) * 0.5px);
    pointer-events: all;
    background: white;
    border-radius: 4px;

    &.justifyLeft {
        right: 50%;
    }
    &.justifyRight {
        left: 50%;
    }
}