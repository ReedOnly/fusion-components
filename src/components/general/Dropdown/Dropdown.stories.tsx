import * as React from 'react';
import { storiesOf } from '@storybook/react';
import Dropdown,{DropdownOption} from './index';
import withFusionStory from '../../../../.storybook/withFusionStory';

const dropdownOptions = [
    {
        key: '1',
        title: 'First',
    },
    {
        key: '2',
        title: 'Selected',
        isSelected: true,
    },
    {
        key: '3',
        title: 'Disabled',
        isDisabled: true,
    },
    {
        key: '4',
        title: 'Last',
    },
];

const DropdownStory = () => {

    const [options, setOptions] = React.useState<DropdownOption[]>(dropdownOptions);
    const [optionsNoLAbel, setOptionsNoLabel] = React.useState<DropdownOption[]>(dropdownOptions);

    const updateOptions = item => options.map(option =>{ return {...option, isSelected : item.key === option.key}});

    return (
        <div style={{ margin: '8px' }}>
            <Dropdown
                options={options}
                label="Dropdown"
                onSelect={item => setOptions(updateOptions(item))}
            />
            <br/>
            <Dropdown
                options={optionsNoLAbel}
                onSelect={item => setOptionsNoLabel(updateOptions(item))}
            />
        </div>
    );
};

storiesOf('General|Dropdown', module)
    .addDecorator(withFusionStory('Dropdown'))
    .add('Default', () => <DropdownStory />);
