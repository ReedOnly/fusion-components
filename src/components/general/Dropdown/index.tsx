import React, { useState, useRef, useCallback, FC } from 'react';
import classNames from 'classnames';
import {
    useClickOutsideOverlayPortal,
    RelativeOverlayPortal,
    useElevationClassName,
} from '@equinor/fusion-components';
import styles from './styles.less';

export type DropdownController = {
    isOpen: boolean;
    setIsOpen: (isOpen: boolean) => void;
    node: React.ReactNode;
    controllerRef: React.MutableRefObject<HTMLElement | null>;
};

type DropdownProps = {
    controller: DropdownController;
};

export const useDropdownController = (
    controller: (
        ref: React.MutableRefObject<HTMLElement | null>,
        isOpen: boolean,
        setIsOpen: (isOpen: boolean) => void
    ) => React.ReactNode
): DropdownController => {
    const [isOpen, setIsOpen] = useState(false);
    const controllerRef = useRef<HTMLElement | null>(null);

    return {
        isOpen,
        setIsOpen,
        node: controller(controllerRef, isOpen, setIsOpen),
        controllerRef,
    };
};

const Dropdown: FC<DropdownProps> = ({ controller, children }) => {
    const { isOpen, setIsOpen, node, controllerRef } = controller;

    const close = useCallback(() => isOpen && setIsOpen(false), [isOpen]);
    useClickOutsideOverlayPortal(close, controllerRef.current);

    const dropdownContainerClassNames = classNames(
        styles.dropdownContainer,
        useElevationClassName(2)
    );

    return (
        <>
            {node}
            <RelativeOverlayPortal relativeRef={controllerRef} show={isOpen}>
                <div className={dropdownContainerClassNames}>{children}</div>
            </RelativeOverlayPortal>
        </>
    );
};

export default Dropdown;
