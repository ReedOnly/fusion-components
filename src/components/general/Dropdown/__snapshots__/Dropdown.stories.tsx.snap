// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General|Dropdown Default 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div
      style={
        Object {
          "margin": "8px",
        }
      }
    >
      <button
        className="container contained primary comfortable"
        disabled={false}
        onClick={[Function]}
        onClickCapture={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        type="button"
      >
        <span
          className="button"
        >
          Show dropdown
        </span>
      </button>
    </div>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;
